File: rovws/src/cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node 
{
public:
    NumberPublisherNode() : Node("number_publisher")    
    {
        this->declare_parameter("number_to_publish" , 2);// aprametre ilanı
        this->declare_parameter("publish_frequency", 1.0);

        number_ = this->get_parameter("number_to_publish").as_int();//parametre atama da böyle yapılıyor
        publish_frequency = this->get_parameter("publish_frequency").as_double();

        number_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        RCLCPP_INFO(this->get_logger(), "number publisher has been started.");
        number_timer_ = this->create_wall_timer(std::chrono::milliseconds((int) (1000.0 / publish_frequency)),
                                         std::bind(&NumberPublisherNode::publishNumber, this));
                                         //periyodu ayarladığımız yerin int olması şartı aranıyor

    }

private:
    
    void publishNumber()
    {
        auto msg = example_interfaces::msg::Int64(); //auto ifadesi msg değişkeninin türünü sağdaki ifadeye bakıp otomatik oalrak belirlenmesini sağlar
        msg.data = number_;
        number_publisher_->publish(msg);
    }

    double publish_frequency;
    int number_;
    rclcpp::TimerBase::SharedPtr number_timer_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr number_publisher_;


    
};


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: rovws/src/py_pkg/py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int32
 
class NumberPublisherNode(Node): 
    def __init__(self):
        super().__init__("number_publisher")
        self.declare_parameter("number_to_publish",10)#aprametre ilanı böyle yapılıyor. 
        #buradaki 2 eğer terminalden bir aprametre girişi yapılmazsa kullanıalcak olan default parametredir
        self.declare_parameter("publish_frequency",1.0)
        self.publish_frequency_ =self.get_parameter("publish_frequency").value
        


        #self.number_ = 2 #burası, aprametre kullanmadan önceki kod parçasıydı
        self.number_=self.get_parameter("number_to_publish").value#aprametreyi yerel değişkene böyle atıyoruz
        self.number_publisher_=self.create_publisher(Int32, "number",10)

        self.number_timer_=self.create_timer(1.0/self.publish_frequency_ , self.publish_number)#1.0/frequency yaptık çünkü o ifade periyodu temsil ediyor.
        


        self.get_logger().info("number publisher has been started")
        #self.publish_number()

    def publish_number(self):
        msg = Int32()
        msg.data = self.number_
        self.number_publisher_.publish(msg)




def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()